File: python/Combination Sum III.py
from typing import List

class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        result = []
        combination = []
        self.backtrack(k, n, 1, combination, result)
        return result

    def backtrack(self, k: int, target: int, start: int, combination: List[int], result: List[List[int]]):
        if target == 0 and len(combination) == k:
            result.append(combination[:])  
            return

        for i in range(start, 10):
            if i > target:
                break
            
            combination.append(i)
            self.backtrack(k, target - i, i + 1, combination, result)
            combination.pop()  

# Test cases
if __name__ == "__main__":
    solution = Solution()
    k = 3
    n = 7
    print(solution.combinationSum3(k, n))  
    
    k = 3
    n = 9
    print(solution.combinationSum3(k, n))  

File: js/Combination Sum III.js
var combinationSum3 = function(k, n) {
    let result = [];
    
    function backtrack(target, start, combination) {
        if (target === 0 && combination.length === k) {
            result.push([...combination]);  
            return;
        }

        for (let i = start; i <= 9; i++) {
            if (i > target) {
                break; 
            }

            combination.push(i);
            backtrack(target - i, i + 1, combination);
            combination.pop(); 
        }
    }
    
    backtrack(n, 1, []);
    return result;
};

console.log(combinationSum3(k = 3, n = 7))
console.log(combinationSum3(k = 3, n = 9))
console.log(combinationSum3(k = 4, n = 1))
